metrics/scaling/k8s_scale_net.sh

```diff
@@ -0,0 +1,281 @@
+
+	local launch_json="$(cat << EOF
+                        "launch_time": {
+                                "Result": $launch_time_ms,
+                                "Units" : "ms"
+                        }
+EOF
+        )"
+        metrics_json_add_array_fragment "$launch_json"
+
+	info "launch [$launch_time_ms]"
+
+	metrics_json_close_array_element
+}
+
+init() {
+        info "Initialising"
+
+	local cmds=("bc" "jq")
+	check_cmds "${cmds[@]}"
+
+	info "Checking Kubernetes accessible"
+	local worked=$( kubectl get nodes > /dev/null 2>&1 && echo $? || echo $? )
+	if [ "$worked" != 0 ]; then
+		die "kubectl failed to get nodes"
+	fi
+
+	info $(get_num_nodes) "Kubernetes nodes in 'Ready' state found"
+
+	k8s_api_init
+
+	 # And now we can set up our results storage then...
+	 metrics_json_init "k8s"
+	 save_config
+}
+
+save_config() {
+	metrics_json_start_array
+
+	local json="$(cat << EOF
+	{
+		"testname": "${TEST_NAME}",
+                "NUM_DEPLOYMENTS": ${NUM_DEPLOYMENTS},
+                "STEP": ${STEP},
+                "wait_time": ${wait_time},
+                "delete_wait_time": ${delete_wait_time},
+                "settle_time": ${settle_time}
+        }
+EOF
+)"
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Config"
+}
+
+run() {
+	info "Running test"
+	local header_post="Content-Type: application/json"
+	local base_curl=${API_ADDRESS}:${API_PORT}/apis/apps/v1/namespaces/default/deployments
+
+	trap cleanup EXIT QUIT KILL
+
+	metrics_json_start_array
+
+	for reqs in $(seq ${STEP} ${STEP} ${NUM_DEPLOYMENTS}); do
+		local deployment="${name_base_depl}${reqs}"
+		info "Testing replicas ${reqs} of ${NUM_DEPLOYMENTS}"
+		# Generate the next yaml file
+
+		local runtime_command
+		if [ -n "$RUNTIME" ]; then
+			runtime_command="s|@RUNTIMECLASS@|${RUNTIME}|g"
+		else
+			runtime_command="/@RUNTIMECLASS@/d"
+		fi
+
+		local input_template
+		local generated_file
+		if [ "$use_api" != "no" ]; then
+			input_template=$input_json
+			generated_file=$generated_json
+		else
+			input_template=$input_yaml
+			generated_file=$generated_yaml
+		fi
+
+		sed -e $runtime_command \
+			-e "s|@DEPLOYMENT@|${deployment}|g" \
+			-e "s|@LABEL@|${LABEL}|g" \
+			-e "s|@LABELVALUE@|${LABELVALUE}|g" \
+			-e "s|@GRACE@|${grace}|g" \
+			< ${input_template} > ${generated_file}
+
+		info "Applying changes"
+		local start_time=$(date +%s%N)
+
+		if [ "$use_api" != "no" ]; then
+			curl -s ${base_curl} -XPOST -H "${header_post}" -d@${generated_file} > /dev/null
+		else
+			kubectl apply -f ${generated_file}
+		fi
+
+		kubectl rollout status --timeout=${wait_time}s deployment/${deployment}
+		kubectl expose deployment $deployment
+
+		set +e
+		while [ $? -ne 0 ]; do
+			kubectl get services $deployment -n default --no-headers=true
+		done
+		set -e
+
+		IP=$(kubectl get services $deployment -n default --no-headers=true | awk '{printf $3}')
+		end_net=$(date +%s%N)
+		info "IP: $IP"
+
+		sleep 3
```

Comment:
@marcemq - I can confirm adding this `sleep 3` here also allows my system to run 20 cycles, and does not hang around the 12th or 13th cycle as before...
This 'fix' should allow us to do the preliminary assessment then of 'does adding more deployments impact the launch/response time of future deployments'. But, if we find it does and we then intend to land this test/PR - we should work out why the 3s delay is necessary and understand that better.

Reply:
@grahamwhaley I've also added `-H "Cache-Control: no-cache"` for the `curl` call.

Can the preliminary assessment be a graph of the reported time metrics? 

> But, if we find it does and we then intend to land this test/PR - we should work out why the 3s delay is necessary and understand that better.

So, the current PR is not going to be considered to be merged? would you please explain more your previous statement?

Reply:
Hi @marcemq - sure, the assessment can be a graph - I like graphs, more than tables of numbers normally. But, @dklyle was saying yesterday - if you just grab 5 sets of numbers say (for 1/10/100/500/1000 pods for instance), then you should be able to tell from just those 5 numbers if the scaling is linear-ish or not... which could be the first level assessment (it may even be easiest to just do those 5 tests by hand)
*If* that looks non-linear, then we could do a finer detailed graph.
*then*, if the graph shows there might be something to chase, we can formalise the PR :-)

For the current PR - I'd not like to merge until we have some clear understanding of why we need the nocache on the curl, and why we need the 3s sleep. If we don't know why, then they are potentially problems hiding that might come back to bite us later. We should understand why we need them.

Reply:
@grahamwhaley the usage of `sleep 3` and `-H "Cache-Control: no-cache"` were removed in the latest commit.

There was a race condition with `curl` call and I've added a service health check to solve it, so the proposed test does not add any noisy metric. The current changes scaled upto 100 deployments and the results are attached in the description box.

Reply:
Great @marcemq - glad that race is found/fixed.
I did a quick graph from your data (of the `time_to_pod_net` result value) - from that 1-100 range, it did not look like the response time was impacted by the number of deployments active - that is - it looked fairly linear (if somewhat noisy).
I suggest you take some further measures of higher numbers of deployments (up to 1000 or 2000) so we can verify things stay linear up to a high number of deployments (quite often we will find there is a 'step function' in speed around the 400-500 pod mark for instance).
You could just run this script with `NUM_DEPLOYMENTS=2000` - but that might take some significant time to run every loop - you may wish to implement the `STEP` ability to launch many deployments in parallel to save time, or you may just wish to do some preliminary tests by hand, before we invest effort in perfecting the script.